More precise typing of the C-- and Mach intermediate languages

xavierleroy opened the pull request:
This branch makes register typing at the C-- and Mach level more precise.

Benefits:
- Safeguard against incorrect optimizations that break the GC interface.
- Possibly smaller sets of roots communicated to the GC.
- More opportunities for common subexpression elimination across allocation points.

What the branch does:

C-- variables and Mach registers are given types, which are currently one of:

`Float`  a unboxed floating-point number
`Int`    an integer or a pointer outside the heap
`Addr`    anything else

The branch splits the `Addr` type into two types, `Val` for well-formed Caml values, `Addr` for derived pointers into the heap.  In other words:
- `Val` denotes a valid OCaml value: either a pointer to the beginning of a heap block, an infix pointer if it is preceded by the correct infix header, or a 2n+1 encoded integer.
- `Int` is for integers (not necessarily 2n+1 encoded) and for pointers outside the heap.
- `Addr` denotes pointers that are neither Val nor Int, i.e. pointers into the heap that point in the middle of a heap block. Such derived pointers are produced by e.g. array indexing.
- `Float` is for unboxed floating-point numbers.

Concerning how variables and registers are tracked by the GC:
- Variables of type `Val` are GC roots. If they are pointers, the GC will not deallocate the addressed heap block, and will update the variable if the heap block moves.
- Variables of type `Int` or `Float` are ignored by the GC. The GC does not change their values.
- Variables of type `Addr` must never be live across an allocation point or function call. (Compiler internal error.)  They cannot be given as roots to the GC because they don't point after a well-formed block header of the kind that the GC needs. However, the GC may move the block pointed into, invalidating the value of the `Addr` variable.

To infer types more precisely, the branch splits the `Word` quantity (type memory_chunk) into `Word_val` (well-formed OCaml values, type `Val`) and `Word_int` (integer but not heap pointer, type `Int`).  Values of type `Addr` are never stored into / loaded from main memory, hence there is no `Word_addr` quantity.

The Cmm language now has no less than three integer addition operators:
- `Caddi`  produces an integer or out-of-heap pointer (type `Int`)
- `Cadda`  produces a derived pointer into the heap (type `Addr`)
- `Caddv`  produces a pointer into the heap that is a well-formed Caml value (type `Val`)

The `Csuba` operator is unused, so I removed it.

The Cmmgen pass is updated to use Word_int/Word_val and Caddi/Cadda/Caddv appropriately.  As a side benefit, some temporary values that point outside the heap (e.g. bigarray payloads) and that used to have type Addr now have the more precise type Int, telling the GC it does not have to track them.

The CSE pass is updated to prevent factorization of subexpressions of type `Addr` across allocation points (this breaks the GC, see PR#6484), but to keep factoring subexpressions of other types across allocation points.  This should result in better CSE of e.g. array check bounds.

The Emit pass is updated with the new cmmtypes and memory_chunks, plus a check that no register of type `Addr` is live across allocation points.

xavierleroy committed:
Experiment: slightly more precise types for C-- variables and Mach registers.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/cmm-mach-types@15567 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

xavierleroy committed:
More precise typing at the C-- and Mach level:
- Register type "Addr" is split into
    . "Val" (well-formed OCaml values, appropriate as GC roots)
    . "Addr" (derived pointers within the heap, must not survive a GC)
- memory_chunk "Word" is split into
    . "Word_val" (OCaml value)
    . "Word_int" (native-sized integer, not a pointer into the heap)

Cmmgen was updated to use Word_val or Word_int as appropriate.

Application #1: fail at compile-time if a derived pointer within the heap
survives a GC point (cf. PR#6484).

Application #2: CSE can do a better job across allocation points
(keep factoring expressions of type Int, Val, Float, but not Addr).


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/cmm-mach-types@15568 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

chambart commented on asmcomp/selectgen.ml:
```diff
@@ -255,7 +256,7 @@ method select_operation op args =
       let (addr, eloc) = self#select_addressing chunk arg1 in
-      if chunk = Word then begin
+      if chunk = Word_int || chunk = Word_val then begin
         let (op, newarg2) = self#select_store true addr arg2 in
```
It could be worth checking that arg2 is effectively of the expected type. That would catch early some kinds of bugs quite complicated to track.

xavierleroy committed:
Don't forget that allocations can run arbitrary Caml code (via signal handlers, etc).  Hence, equations involving loads must be erased across allocations.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/cmm-mach-types@15569 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

xavierleroy committed:
Prudently revert the type of "Cconst_symbol" to typ_val instead of typ_int.
There seems to be issues with let-bound variables initialized with symbol
addresses and later updated.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/cmm-mach-types@15572 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

alainfrisch commented:
> If this pattern is forbidden it should be explicitly documented.

I can only confirm that I consider this an important property, since it could allow lifting bound checks (and some indirection) out of loops, or removing some of them as in:

```
   a.{i} <- f a.{i}
```

gasche commented:
During the last developer meeting it was decided that the patch is mergeable. @xavierleroy , will you merge it yourself?

xavierleroy committed:
Update to trunk r16267.
A change on trunk in cmmgen was causing problems: the use of array_indexing
in bigarray_indexing was constructing derived pointers of Addr type
instead of the correct Int type.  (Because bigarray inner pointers
are outside the heap, and because they are live across allocations.)
Added an optional argument to array_indexing to specify expected
pointer type, and adapted bigarray_indexing to use it with type Int.


git-svn-id: http://caml.inria.fr/svn/ocaml/branches/cmm-mach-types@16268 f963ae5c-01c2-4b8c-9fe0-0dff7051ff02

xavierleroy commented:
Updated this branch to the current trunk.  There was a problem in Cmmgen.bigarray_indexing.  I'll let Travis re-do the testing before final merge.

xavierleroy commented:
Travis is happy.  Merged in trunk (r16269), will be part of 4.03.  Closing this PR.
